File: backend/app/schemas/scrape.py
from datetime import datetime
from typing import Any
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, HttpUrl

from app.models.enums import ScrapeJobStatus, ScrapeResultStatus, ScrapeSourceType


class ScrapeJobCreate(BaseModel):
    name: str = Field(min_length=2, max_length=255)
    description: str | None = None
    source_type: ScrapeSourceType
    target_url: HttpUrl
    config: dict[str, Any] = Field(default_factory=dict)
    schedule_cron: str | None = Field(default=None, max_length=64)


class ScrapeJobUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=2, max_length=255)
    description: str | None = None
    source_type: ScrapeSourceType | None = None
    target_url: HttpUrl | None = None
    config: dict[str, Any] | None = None
    schedule_cron: str | None = Field(default=None, max_length=64)
    is_active: bool | None = None
    status: ScrapeJobStatus | None = None


class ScrapeJobResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    organization_id: UUID
    created_by_id: UUID | None
    name: str
    description: str | None
    source_type: ScrapeSourceType
    target_url: str
    config: dict[str, Any]
    schedule_cron: str | None
    is_active: bool
    status: ScrapeJobStatus
    last_run_at: datetime | None
    created_at: datetime
    updated_at: datetime


class ScrapeResultResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True, populate_by_name=True)

    id: UUID
    job_id: UUID
    triggered_by_id: UUID | None
    status: ScrapeResultStatus
    started_at: datetime | None
    finished_at: datetime | None
    page_title: str | None
    raw_html: str | None
    extracted_data: dict[str, Any] | None
    items_count: int | None
    summary: str | None
    error_message: str | None
    artifact_urls: dict[str, Any] | None
    metadata: dict[str, Any] | None = Field(validation_alias="metadata_")
    created_at: datetime
    updated_at: datetime


class ScrapeRunResponse(BaseModel):
    result_id: UUID
    job_id: UUID
    status: ScrapeResultStatus
    message: str


class PaginatedScrapeJobsResponse(BaseModel):
    items: list[ScrapeJobResponse]
    total: int
    limit: int
    offset: int


class PaginatedScrapeResultsResponse(BaseModel):
    items: list[ScrapeResultResponse]
    total: int
    limit: int
    offset: int

File: frontend/src/types/scrape.ts
import type { ScrapeSourceType } from '@/types/signal'

export type ScrapeJobStatus = 'draft' | 'active' | 'paused' | 'archived'

export type ScrapeResultStatus = 'pending' | 'running' | 'success' | 'failed' | 'cancelled'

export interface ScrapeJob {
  id: string
  organization_id: string
  created_by_id: string | null
  name: string
  description: string | null
  source_type: ScrapeSourceType
  target_url: string
  config: Record<string, unknown>
  schedule_cron: string | null
  is_active: boolean
  status: ScrapeJobStatus
  last_run_at: string | null
  created_at: string
  updated_at: string
}

export interface ScrapeResult {
  id: string
  job_id: string
  triggered_by_id: string | null
  status: ScrapeResultStatus
  started_at: string | null
  finished_at: string | null
  page_title: string | null
  raw_html: string | null
  extracted_data: Record<string, unknown> | null
  items_count: number | null
  summary: string | null
  error_message: string | null
  artifact_urls: Record<string, unknown> | null
  metadata: Record<string, unknown> | null
  created_at: string
  updated_at: string
}

export interface PaginatedScrapeJobsResponse {
  items: ScrapeJob[]
  total: number
  limit: number
  offset: number
}

export interface PaginatedScrapeResultsResponse {
  items: ScrapeResult[]
  total: number
  limit: number
  offset: number
}

export interface ScrapeRunResponse {
  result_id: string
  job_id: string
  status: ScrapeResultStatus
  message: string
}

export interface ScrapeJobCreateRequest {
  name: string
  description?: string | null
  source_type: ScrapeSourceType
  target_url: string
  schedule_cron?: string | null
}

export interface ScrapeJobListParams {
  status?: ScrapeJobStatus
  source_type?: ScrapeSourceType
  is_active?: boolean
  limit?: number
  offset?: number
}

export type ScrapeJobSortField =
  | 'name'
  | 'source_type'
  | 'status'
  | 'target_url'
  | 'last_run_at'
  | 'created_at'

export type SortDirection = 'asc' | 'desc'
